### PR TITLE
Implement locality aware load balancing

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/ztunnel/issues/862

### CODE REVIEW

File: proto/workload.proto [hunk 1]
```diff
@@ -74,6 +74,36 @@ message Service {
...
+    UNSPECIFIED = 0;
+    REGION = 1;
+    ZONE = 2;
+    SUBZONE = 3;
+    NODE = 4;
+    CLUSTER = 5;
+    NETWORK = 6;
+  }
+  enum Mode {
+    UNSPECIFIEDx = 0;
```
Comment (louiscryan) [inline proto/workload.proto:95]: typo?

File: proto/workload.proto [hunk 2]
```diff
@@ -74,6 +74,36 @@ message Service {
...
+
+    STRICT = 1;
+    FAILOVER = 2;
+
+  }
+  // routing_preference defines what scopes we want to keep traffic within. These are ANDed together
+  // For example [LOCALITY, NETWORK] means we want to keep traffic within the same locality and network.
```
Comment (louiscryan) [inline proto/workload.proto:102]: no 'LOCALITY' value

File: proto/workload.proto [hunk 3]
```diff
@@ -74,6 +74,36 @@ message Service {
...
   GatewayAddress waypoint = 7;
+
+  // Load balancing policy for selecting endpoints.
+  // Note: this applies only to connecting directly to the workload; when waypoints are used, the waypoint's load_balancing
+  // configuration is used.
+  LoadBalancing load_balancing = 8;
+}
+
+message LoadBalancing {
+  enum Targets {
```
Comment (louiscryan) [inline proto/workload.proto:85]: Target is a pretty vague term, maybe 'Locality'

File: proto/workload.proto [hunk 4]
```diff
@@ -74,6 +74,36 @@ message Service {
...
+    UNSPECIFIED = 0;
```
Comment (louiscryan) [inline proto/workload.proto:86]: would ANY or NONE read more cleanly in docs.

"When the locality_preference is set to [NONE] traffic can be routed to any endpoint that is a member of the mesh"

File: proto/workload.proto [hunk 5]
```diff
@@ -74,6 +74,36 @@ message Service {
...
+    ZONE = 2;
+    SUBZONE = 3;
+    NODE = 4;
+    CLUSTER = 5;
+    NETWORK = 6;
+  }
+  enum Mode {
+    UNSPECIFIEDx = 0;
+
+    STRICT = 1;
```
Comment (louiscryan) [inline proto/workload.proto:97]: Doc.  "Only if the set of endpoints in each target defined in routing_preference is empty or every endpoint is unreachable can an endpoint in the next target in routing_preference be used"

File: src/state.rs [hunk 6]
```diff
@@ -250,6 +255,60 @@ impl ProxyState {
         }
         None
     }
+
+    fn load_balance<'a>(&self, src: &Workload, svc: &'a Service) -> Option<&'a Endpoint> {
```
Comment (ilrudie) [inline src/state.rs:259]: this seems to ignore failover/strict and always does a best match, yeah?

Comment (howardjohn) [reply src/state.rs:259]:
```
                        // Doesn't match all, and required to. Do not select this endpoint
                        if lb.mode == LoadBalancerMode::Strict && rank != lb.targets.len() {
                            return None;
                        }
```

should do it

File: src/state.rs [hunk 7]
```diff
@@ -250,6 +255,62 @@ impl ProxyState {
...
+                                    src.locality.subzone == wl.locality.subzone
+                                }
+                                LoadBalancerScopes::Node => src.node == wl.node,
+                                LoadBalancerScopes::Cluster => src.cluster_id == wl.cluster_id,
+                                LoadBalancerScopes::Network => src.network == wl.network,
+                            };
+                            if matches {
+                                rank += 1;
+                            } else {
+                                break;
```
Comment (stevenctl) [inline src/state.rs:293]: We should document if we always expect the scopes to be "size-ordered"

Comment (howardjohn) [reply src/state.rs:293]: Its meant to be documented here:
```
   // In FAILOVER mode, endpoint selection will prefer endpoints that match all preferences, but failover to groups of endpoints
    // that match less (or, eventually, none) preferences.
    // For instance, with `[NETWORK, REGION, ZONE]`, we will send to:
    // 1. Endpoints matching `[NETWORK, REGION, ZONE]`
    // 2. Endpoints matching `[NETWORK, REGION]`
    // 3. Endpoints matching `[NETWORK]`
    // 4. Any endpoints
```

Which is also how failover works in Istio-envoy

File: src/state.rs [hunk 8]
```diff
@@ -575,8 +644,8 @@ impl DemandProxyState {
             }
         };
         let wp_socket_addr = SocketAddr::new(wp_nw_addr.address, gw_address.hbone_mtls_port);
-        match self
-            .fetch_upstream(&wp_nw_addr.network, wp_socket_addr)
+        match self // THIS IS THE WRONG WORKLOAD
```
Comment (stevenctl) [inline src/state.rs:647]: just need to pass in src, right?

**Comment by stevenctl** [comment]
LGTM

File: src/state.rs (hunk 7 above)
Comment (ilrudie) [reply src/state.rs:293]: I was also going to ask about this wrt the `break` being ok or not. I think it relies on the routing preferences being guaranteed ordered correctly.

File: src/state.rs (hunk 8 above)
Comment (howardjohn) [reply src/state.rs:647]: ah good catch, forgot about this todo. Yep

File: src/state.rs (hunk 7 above)
Comment (howardjohn) [reply src/state.rs:293]: My intended implementation was to make order matter here. Open to changes though

**Comment by howardjohn** [comment]
@stevenctl this should be good to go, will open istio side once this lands

File: src/state.rs (hunk 7 above)
Comment (ilrudie) [reply src/state.rs:293]: I think that's fine as long as folks are aware that it matters.

**Review by stevenctl — APPROVED** [review]

File: proto/workload.proto [hunk 9]
```diff
@@ -74,6 +74,52 @@ message Service {
...
+    // 3. Endpoints matching `[NETWORK]`
+    // 4. Any endpoints
+    FAILOVER = 2;
+  }
+
+  // routing_preference defines what scopes we want to keep traffic within.
+  // The `mode` determines how these routing preferences are handled
+  repeated Scope routing_preference = 1;
+  // mode defines how we should handle the routing preferences.
+  Mode mode = 2;
```
Comment (hzxuzhonghu) [inline proto/workload.proto:122]: With this api, how can user set it for `strict`? 

And how can we expand to support other algorithms with this struct?

Comment (daixiang0) [reply proto/workload.proto:122]: Package this as one algorithm if need to expand?

**Pull request merged**